### PR TITLE
Pubsub/backend dep injection

### PR DESCRIPTION
**Pull request by NelsonVides**
This PR addresses GDPR pubsub removal, including the refactor of `mod_pubsub_db_backend`

### CODE REVIEW

File: src/pubsub/mod_pubsub.erl [hunk 1]
```diff
@@ -63,6 +63,7 @@
 -define(STDNODE_MODULE, node_flat).
 -define(PEPNODE, <<"pep">>).
 -define(PUSHNODE, <<"push">>).
+-define(NO_DEFAULT_CACHE, no_last_item_cache).
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: why actually do we need to define that macro ? Can't we use `no_last_item_cache` directly ?

File: src/pubsub/mod_pubsub.erl [hunk 2]
```diff
@@ -822,6 +861,7 @@ out_subscription(Acc, User, Server, JID, subscribed) ->
 out_subscription(Acc, _, _, _, _) ->
     Acc.
 
+%% EXPORTED FOR REAL
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: what does it mean ?

File: src/pubsub/mod_pubsub.erl [hunk 3]
```diff
@@ -830,37 +870,39 @@ out_subscription(Acc, _, _, _, _) ->
...
+unsubscribe_user(Backend, Host, Entity, Owner) ->
     BJID = jid:to_lower(jid:to_bare(Owner)),
     lists:foreach(fun (PType) ->
-                         unsubscribe_user_per_plugin(Host, Entity, BJID, PType)
+                         unsubscribe_user_per_plugin(Backend, Host, Entity, BJID, PType)
                   end, plugins(Host)).
 
-unsubscribe_user_per_plugin(Host, Entity, BJID, PType) ->
-    {result, Subs} = node_action(Host, PType, get_entity_subscriptions, [Host, Entity]),
+%% TODO: duplicated logic maybe?
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: could you elaborate more on that?

File: src/pubsub/nodetree_dag.erl [hunk 4]
```diff
@@ -132,26 +132,27 @@ remove_config_parent(Node, [H | T], Acc) ->
     remove_config_parent(Node, T, [H | Acc]).
 
 -spec validate_parentage(
+        Backend :: atom(),
```
Comment (ludwikbukowski) [inline src/pubsub/nodetree_dag.erl]: there is built-in type `module()`

File: src/pubsub/mod_pubsub.erl [hunk 5]
```diff
@@ -198,7 +199,9 @@
           max_subscriptions_node  = undefined,
           default_node_config     = [],
           nodetree                = <<"nodetree_", (?STDTREE)/binary>>,
-          plugins                 = [?STDNODE]
+          plugins                 = [?STDNODE],
+          backend                 = mnesia,
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: actually for that case, I would use macro like `DEFAULT_BACKEND` in the beginning of the module

File: src/pubsub/mod_pubsub.erl [hunk 6]
```diff
@@ -349,6 +365,10 @@ store_config_in_ets(Host, ServerHost, Opts, Plugins, NodeTree, PepMapping) ->
                                      fun(A) when is_list(A) -> filter_node_options(A) end, []),
     ItemPublisher = gen_mod:get_opt(item_publisher, Opts,
                                     fun(A) when is_boolean(A) -> A end, false),
+    ProxyBackend = gen_mod:get_opt(proxy_backend, Opts, fun(A) when is_atom(A) -> A end,
+                                   mod_pubsub_db_backend),
+    ProxyCache = gen_mod:get_opt(proxy_cache, Opts, fun(A) when is_atom(A) -> A end,
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: seems that you are introducing new options to the pubsub module, can you document them?

File: src/pubsub/mod_pubsub.erl [hunk 7]
```diff
@@ -349,6 +365,10 @@ store_config_in_ets(Host, ServerHost, Opts, Plugins, NodeTree, PepMapping) ->
                                      fun(A) when is_list(A) -> filter_node_options(A) end, []),
     ItemPublisher = gen_mod:get_opt(item_publisher, Opts,
                                     fun(A) when is_boolean(A) -> A end, false),
+    ProxyBackend = gen_mod:get_opt(proxy_backend, Opts, fun(A) when is_atom(A) -> A end,
```
Comment (ludwikbukowski) [inline src/pubsub/mod_pubsub.erl]: I suppose that could be created based on `backend` option ?

**Comment by mongoose-im** [comment]
[6732.1](https://travis-ci.org/esl/MongooseIM/jobs/550322912) / Erlang 20.3 / small_tests / 964c0374aefe689f28ee0c741fe8826574520d7a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2348/6732/small_tests.20.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2348/6732/small_tests.20.3/small/ct_run.mongooseim@localhost.2019-06-25_14.32.56/lib.mongooseim.logs/run.2019-06-25_14.32.57/suite.log.html)

File: src/pubsub/mod_pubsub.erl (hunk 1 above)
Comment (NelsonVides) [reply src/pubsub/mod_pubsub.erl]: I was using that same atom in a couple of places already, dunno, should I remove the macro?

File: src/pubsub/mod_pubsub.erl (hunk 2 above)
Comment (NelsonVides) [reply src/pubsub/mod_pubsub.erl]: this is me committing too much. I was just telling myself which functions are really part of the external API and I shouldn't touch

File: src/pubsub/mod_pubsub.erl (hunk 3 above)
Comment (NelsonVides) [reply src/pubsub/mod_pubsub.erl]: that is code for the exported function `unsubscribe_user`, but we wrote a very similar functionality for our `remove_user`, where the only difference is basically whether backend is loaded or we're trying with all of them. I thought we could review that.

File: src/pubsub/nodetree_dag.erl (hunk 4 above)
Comment (NelsonVides) [reply src/pubsub/nodetree_dag.erl]: ah, didn't know that type!

File: src/pubsub/mod_pubsub.erl (hunk 2 above)
Comment (ludwikbukowski) [reply src/pubsub/mod_pubsub.erl]: Ok, can we remove it ?

File: src/pubsub/mod_pubsub.erl (hunk 3 above)
Comment (ludwikbukowski) [reply src/pubsub/mod_pubsub.erl]: yeah, if you can reuse existing functions, lets do that

Comment (NelsonVides) [reply src/pubsub/mod_pubsub.erl]: But this requires a bit more rework I didn't have the time to do in this effort, can you help me out with that one?

**Comment by mongoose-im** [comment]
[6734.1](https://travis-ci.org/esl/MongooseIM/jobs/550342550) / Erlang 20.3 / small_tests / a59c55bf558db90bd7f1aae2548b1c3f43c9506c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2348/6734/small_tests.20.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2348/6734/small_tests.20.3/small/ct_run.mongooseim@localhost.2019-06-25_17.44.23/lib.mongooseim.logs/run.2019-06-25_17.44.24/suite.log.html)

**Pull request closed without merging**